### PR TITLE
Provide service connection support for Hazelcast

### PR DESCRIPTION
**Pull request by nosan**
gh-34778

### CODE REVIEW

**Comment by nosan** [comment]
The open question is about `hazelcast/hazelcast-enterprise` image, should I add it as well?

File: spring-boot-project/spring-boot-docker-compose/src/main/java/org/springframework/boot/docker/compose/service/connection/hazelcast/HazelcastDockerComposeConnectionDetailsFactory.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+		private final String host;
+
+		private final int port;
+
+		private final String clusterName;
+
+		HazelcastDockerComposeConnectionDetails(RunningService service) {
+			super(service);
+			this.clusterName = service.env().get(CLUSTER_NAME_ENV);
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-docker-compose/src/main/java/org/springframework/boot/docker/compose/service/connection/hazelcast/HazelcastDockerComposeConnectionDetailsFactory.java:65]: All our existing `DockerComposeConnectionDetails` use a `...Environment` class to handle the environment variables, see for example `MongoEnvironment` in the `MongoDockerComposeConnectionDetailsFactory`. I think we should do that here, too.

File: spring-boot-project/spring-boot-docker-compose/src/main/resources/META-INF/spring.factories [hunk 2]
```diff
@@ -33,4 +33,5 @@ org.springframework.boot.docker.compose.service.connection.rabbit.RabbitDockerCo
 org.springframework.boot.docker.compose.service.connection.redis.RedisDockerComposeConnectionDetailsFactory,\
 org.springframework.boot.docker.compose.service.connection.sqlserver.SqlServerJdbcDockerComposeConnectionDetailsFactory,\
 org.springframework.boot.docker.compose.service.connection.sqlserver.SqlServerR2dbcDockerComposeConnectionDetailsFactory,\
-org.springframework.boot.docker.compose.service.connection.zipkin.ZipkinDockerComposeConnectionDetailsFactory
+org.springframework.boot.docker.compose.service.connection.zipkin.ZipkinDockerComposeConnectionDetailsFactory,\
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-docker-compose/src/main/resources/META-INF/spring.factories:36]: The entries in that file should be sorted alphabetically.

File: spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/features/dev-services.adoc [hunk 3]
```diff
@@ -130,6 +130,9 @@ The following service connections are currently supported:
 
 | `ZipkinConnectionDetails`
 | Containers named "openzipkin/zipkin".
+
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/features/dev-services.adoc:133]: The entries in that table should be sorted alphabetically.

File: spring-boot-project/spring-boot-testcontainers/build.gradle [hunk 4]
```diff
@@ -51,6 +51,7 @@ dependencies {
 	dockerTestImplementation("org.springframework.ldap:spring-ldap-core")
 	dockerTestImplementation("org.springframework.pulsar:spring-pulsar")
 	dockerTestImplementation("org.testcontainers:junit-jupiter")
+	dockerTestImplementation("com.hazelcast:hazelcast")
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-testcontainers/build.gradle:54]: The `dockerTestImplementation` declarations should be sorted alphabetically.

File: spring-boot-project/spring-boot-testcontainers/src/main/resources/META-INF/spring.factories [hunk 5]
```diff
@@ -38,4 +38,5 @@ org.springframework.boot.testcontainers.service.connection.r2dbc.PostgresR2dbcCo
 org.springframework.boot.testcontainers.service.connection.r2dbc.SqlServerR2dbcContainerConnectionDetailsFactory,\
 org.springframework.boot.testcontainers.service.connection.redis.RedisContainerConnectionDetailsFactory,\
 org.springframework.boot.testcontainers.service.connection.redpanda.RedpandaContainerConnectionDetailsFactory,\
-org.springframework.boot.testcontainers.service.connection.zipkin.ZipkinContainerConnectionDetailsFactory
+org.springframework.boot.testcontainers.service.connection.zipkin.ZipkinContainerConnectionDetailsFactory,\
+org.springframework.boot.testcontainers.service.connection.hazelcast.HazelcastContainerConnectionDetailsFactory
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-testcontainers/src/main/resources/META-INF/spring.factories:42]: The entries in that file should be sorted alphabetically.

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java [hunk 6]
```diff
@@ -272,7 +272,12 @@ public enum TestImage {
 	/**
 	 * A container image suitable for testing Redis via Bitnami.
 	 */
-	BITNAMI_REDIS("bitnami/redis", "7.2.4");
+	BITNAMI_REDIS("bitnami/redis", "7.2.4"),
+
+	/**
+	 * A container image suitable for testing Hazelcast.
+	 */
+	HAZELCAST("hazelcast/hazelcast", "5.5.0-slim", () -> HazelcastContainer.class);
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java:280]: The values in that enum should be sorted alphabetically. (All the bitnami images are at the bottom).

**Review by mhalbritter — COMMENTED** [review]
Hey @nosan, thanks for the PR! I've left some comments for your consideration.

**Comment by nosan** [comment]
> I think I don't understand this sentence. Does that mean I don't have to set the license key on the client which connects to the enterprise server?

Yes, you are right. You don't need to configure a client with a license key to connect to `hazelcast-enterprise`.

>I can't start hazelcast/hazelcast-enterprise without setting an enterprise key.

I verified this locally (I have a license key).

I am asking because I understand that adding something without adding a test for this image does not sound right, or I can add it but in that case, LICENSE_KEY should be configured on Spring Boot CI with some JUnit conditionals and of course, you need a license key.

**Comment by mhalbritter** [comment]
Thanks for the clarification. I'll talk to the team about this.

**Comment by wilkinsona** [comment]
I don't think we should add support for `hazelcast-enterprise` as we can't (easily) test it.

File: spring-boot-project/spring-boot-testcontainers/src/main/resources/META-INF/spring.factories (hunk 5 above)
Comment (nosan) [reply spring-boot-project/spring-boot-testcontainers/src/main/resources/META-INF/spring.factories:42]: Sorted by qualified name or just class name?

**Comment by nosan** [comment]
Thank you very much for your review @mhalbritter 
I fixed your comments. I hope I did not mess with sorting.

**Comment by mhalbritter** [comment]
Thanks @nosan !

**Comment by nosan** [comment]
Thank you @mhalbritter

**Pull request closed without merging**